CheckRebootHygiene Actor

Wecros opened the pull request:
OAMG-8806

DONE:
- uptime check
- mismatched kernel versions check
- files modified in /boot since last reboot check
- tests for all checks

**Repo structure**

Custom actors are located under `"repos/system_upgrade_supplements/{common, el7toel8, el8toel9}"`. The `common`, `el7toel8`, and `el8toel9` are leapp repositories (.leapp files). They are linked to the respective repositories in [leapp-repository](https://github.com/oamg/leapp-repository) (under `system_upgrade`). Additionally, `el7toel8` and `el8toel9` repos are linked to the `common` supplements repository.


Questions
- should uptime have its own model and actor? Similarly for kernel versions + boot file
  - not needed / should not be done for this repository. We want to install individual actors so introducing actor dependencies would make the process harder.

Wecros committed:
Add CheckRebootHygiene actor

The CheckRebootHygiene actor will report an inhibitor in case of conditions
that may be prohibited by an organization's reboot hygiene policy.

The actor will report inhibitor risk when:
* The host uptime is greater than the maximum defined by the policy.
* The running kernel version does not match the default kernel version
  configured in the bootloader.
* Any files are found under /boot that have been modified since the last reboot.

Set repository structure for leapp-supplements:
* Actors are placed under repos/system_upgrade_common folder. In it
  there is a common, el7toel8, el8toel9 subdirectories.
* supplements leapp repositories are linked to the respective
  repositories in leapp-repository and to the common supplements
  repository.

Wecros commented:
Output from rhel7 machine, adjusted the uptime for 0 days check.
```
[ec2-user@ip-172-31-22-169 ~]$ snactor run --debug --print-output CheckUptime

[
  {
    "stamp": "2023-04-13T06:57:38.860727Z",
    "hostname": "ip-172-31-22-169.eu-central-1.compute.internal",
    "actor": "check_uptime",
    "topic": "report_topic",
    "context": "b2054a9b-3dba-4e94-b7aa-1a813dbc0c50",
    "phase": "NON-WORKFLOW-EXECUTION",
    "message": {
      "hash": "a8a870070d33e3c52df991beda183ecb7aef8125cfff038ad819b847edbdd8e8",
      "data": "{\"report\": \"{\\\"audience\\\": \\\"sysadmin\\\", \\\"flags\\\": [\\\"inhibitor\\\"], \\\"groups\\\": [\\\"sanity\\\", \\\"inhibitor\\\"], \\\"key\\\": \\\"25b62d16befc450cef2a240fc7c385f42b2b73d6\\\", \\\"severity\\\": \\\"high\\\", \\\"summary\\\": \\\"Excessive uptime detected: 0 days. Consider rebooting the system before proceeding with the upgrade.\\\", \\\"title\\\": \\\"Excessive uptime detected\\\"}\"}"
    },
    "type": "Report"
  }
]
```

pirat89 commented on repos/.leapp/info:
```diff
@@ -0,0 +1 @@
+{"name": "repos", "id": "411897a6-6503-46cd-be5b-b29c5154e0c0"}
```
the name repos does not say so much. Imagine that someone else use the same name of the repo. so there will be a collision. regarding the git reponame, i would expect something like
* `system_upgrade_supplements`

Also, do not have a repo in directory "repos". The directory suggests I should be able to find more repositories inside possibly in fugure. put the directory into a separate dir, e.g. "repos/upgrade/supplementary".

Also the repository should have set dependency on the system_upgrade_common repository, which is not set currently based on this.

Wecros replied:
Now I understand the structure of leapp-repositry, thanks for the correcting.

I created `system_upgrade/supplements` with repo link to `system_upgrade_common`

pirat89 commented on repos/actors/checkuptime/libraries/checkuptime.py:
```diff
@@ -0,0 +1,27 @@
+        uptime_seconds = float(uptime_file.readline().split()[0])
+
+    if uptime_seconds > excessive_uptime_limit:
+        summary = (
+            "Excessive uptime detected: {:.0f} days. "
+            "Consider rebooting the system before proceeding with the upgrade."
+        ).format(uptime_seconds / (24 * 60 * 60))
+
+        report = [
+            reporting.Title("Excessive uptime detected"),
+            reporting.Summary(summary),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Groups([reporting.Groups.SANITY, reporting.Groups.INHIBITOR]),
+        ]
```
The report seems to me weird.
* I am missing explanation why this should be problem. The summary is actualy leaving me in blind (I consider this pretty common to see a system that is running for longer than 1month). 
* Also reading the summary, I would understand the msg as purely informative, however then I will realize i has high severity and it's inhibitor so it blocks the upgrade
* From the msg, I do not have idea what is actually the excessive uptime. Where is the limit? How this limit has been set?
* `Consider rebooting the system before proceeding with the upgrade.` is not summary, it's remediation hint. In this case it's written in the way that people could do it if they want. But the inhibitor says they must do it.

Wecros replied:
Thanks for the feedback! The previous commit just outlined the report. I incorporated [Bob's comment](https://github.com/oamg/leapp-supplements/pull/1#issuecomment-1506887106) into the report messages now.

I listed the mismatched kernel versions and modified /boot files as related sources instead of putting them into the entry summary. Is that expected?

pirat89 commented on repos/actors/checkuptime/libraries/checkuptime.py:
```diff
@@ -0,0 +1,27 @@
+
+def check_excessive_uptime():
+    # TODO: can this be parametrized?
+    excessive_uptime_limit = 30 * 24 * 60 * 60  # 30 days in seconds
```
define constants. from the code I see 2
* DAY_IN_SECONDS
* EXCESSIVE_UPTIME_LIMIT

Wecros replied:
done

pirat89 commented on repos/actors/checkuptime/libraries/checkuptime.py:
```diff
@@ -0,0 +1,27 @@
+
+
+def check_excessive_uptime():
+    # TODO: can this be parametrized?
```
how the parametrization helps here? currently there is not a configuration for actors, and if user can parametrize it, how this is helpful?

Wecros replied:
I think it could be beneficial for the customers to have the excessive uptime parameterized when building the repository RPM via some Makefile parameterization (e.g., env var with default value of 30 days).

Is this something that would be beneficial to customers? @swapdisk Or is hardcoded value of 30 days enough for this use case.
Anyway this would be done as part of OAMG-8807.

swapdisk replied:
Makefile parameterization (e.g., env var with default value of 30 days) would be nice. That way, we define a model for how parameterization can be done for future actors that may also benefit customers wanting to customize things. Makefile variables could be defined to customize which actors would be included in the package build.

pirat89 commented on repos/tags/checks.py:
```diff
@@ -0,0 +1,9 @@
+
+
+class TargetTransactionChecksPhaseTag(Tag):
+    name = 'target_checks'
```
remove. you cannot redefine existing tags

pirat89 commented on repos/tags/ipu.py:
```diff
@@ -0,0 +1,5 @@
+from leapp.tags import Tag
+
+
+class IPUWorkflowTag(Tag):
+    name = 'ipu'
```
remove, you cannot redefine existing tags

pirat89 commented on repos/actors/checkuptime/actor.py:
```diff
@@ -0,0 +1,19 @@
+    name = 'check_uptime'
+    consumes = ()
+    produces = (Report,)
+    tags = (TargetTransactionChecksPhaseTag, IPUWorkflowTag)
```
Executing the actor in wrong phase. what this actor has common with the upgrade (dnf) transaction? from that, I would expect 2 actors - scanner and checker in first two phases

Wecros replied:
Yes, I see that these two were not put there well. I removed them in favor of `ChecksPhaseTag`. I am still confused which tags to use in which situations though. Can you please give me a hand here?

pirat89 replied:
sorry for the delayed response. One actor solution is good enough regarding the context of this repository. Just switch it to the `ChecksPhaseTag` instead. Just ping me when you are online during the day for a better guidance. Or you can just follow the IPUWorkflow definition: https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/common/workflows/inplace_upgrade.py

swapdisk commented:
@Wecros and @pirat89 , thank you for working on this PR. Here's some example verbiage adapted from an OpenSCAP check I developed for these requirements. You can parse these to pad out the texts for rationale, summary, remediation hint, etc. The idea is that an organization would have a policy that OS maintenance that will require a reboot shall not be undertaken if the uptime is greater than some threshold. 

> For the README: The CheckRebootHygiene actor will report an inhibitor in case of conditions that may be prohibited by an organization's reboot hygiene policy. For example, the actor will report inhibitor risk if the host uptime is greater than a maximum defined by the policy, if the running kernel version does not match the default kernel version configured in the bootloader, or if any files are found under /boot that have been modified since the last reboot.

> This host has not been rebooted for over {{max_allowed_uptime_text}}. To reduce the risk of boot issues related to any changes made since the last reboot, policy requires the host to be rebooted before going forward with the upgrade. 

> This host is not running the default kernel version currently configured in the bootloader. Maybe a new kernel version was installed without then rebooting the host? Policy requires the host to be rebooted before going forward with the upgrade.
>
> Bootloader default: {{default_kernel_version}}
> Running kernel: {{uname_r_version}}

> Files on the /boot partition listed below have been modified since the last reboot. To reduce the risk of boot issues related to changes made since the last reboot, policy requires the host to be rebooted before going forward with the upgrade. 
>
> {{list_of_files_modified}}

Wecros commented:
Thinking whether the reboot hygiene checks should be all in one actor. Or whether the capability of getting (scanning) for current uptime, current and default kernel versions, and /boot files should not be split to different actors - which would produce this information - and CheckRebootHygiene actor would then only consume this info via leapp api. 

What is the standard way of doing things in these scenarios? @pirat89

Wecros commented on repos/system_upgrade/supplements/topics/systeminfo.py:
```diff
@@ -0,0 +1,5 @@
+from leapp.topics import Topic
+
+
+class SystemInfoTopic(Topic):
```
Remove this because already defined in leapp-repository?

pirat89 replied:
yep, remove

Wecros commented on repos/system_upgrade/supplements/actors/checkreboothygiene/libraries/checkreboothygiene.py:
```diff
@@ -0,0 +1,126 @@
+
+# NOTE: got from repos/system_upgrade/common/actors/ipuworkflowconfig/libraries/ipuworkflowconfig.py
+#       should this be moved to a common library?
+def _get_booted_kernel():
```
Move this to common library (used by ipuworkflowconfig actor)? Or create actor which produces the booted kernel version...

pirat89 replied:
> @Wecros and @pirat89 , thank you for working on this PR. Here's some example verbiage adapted from an OpenSCAP check I developed for these requirements. You can parse these to pad out the texts for rationale, summary, remediation hint, etc. The idea is that an organization would have a policy that OS maintenance that will require a reboot shall not be undertaken if the uptime is greater than some threshold.
> 
> > For the README: The CheckRebootHygiene actor will report an inhibitor in case of conditions that may be prohibited by an organization's reboot hygiene policy. For example, the actor will report inhibitor risk if the host uptime is greater than a maximum defined by the policy, if the running kernel version does not match the default kernel version configured in the bootloader, or if any files are found under /boot that have been modified since the last reboot.
> 
> > This host has not been rebooted for over {{max_allowed_uptime_text}}. To reduce the risk of boot issues related to any changes made since the last reboot, policy requires the host to be rebooted before going forward with the upgrade.
> 
> > This host is not running the default kernel version currently configured in the bootloader. Maybe a new kernel version was installed without then rebooting the host? Policy requires the host to be rebooted before going forward with the upgrade.
> > Bootloader default: {{default_kernel_version}}
> > Running kernel: {{uname_r_version}}
> 
> > Files on the /boot partition listed below have been modified since the last reboot. To reduce the risk of boot issues related to changes made since the last reboot, policy requires the host to be rebooted before going forward with the upgrade.
> > {{list_of_files_modified}}

Thanks for providing more info. Something like that is good to have in the description of the actor so people know exactly what to expect from the actor and whether they want to apply it.

pirat89 replied:
> Move this to common library (used by ipuworkflowconfig actor)? Or create actor which produces the booted kernel version...

let's discuss that after the sync mtg.

Wecros commented:
Updated the PR with following changes:
* Removed previously defined tags
* Restructured the repository so that it depends on  system_ugprade_common and can be found under system_upgrade/supplements
* Renamed checkuptime to checkreboothygiene
* Implemented mismatched kernel versions check.
* Implemented changed /boot files since last boot check.
* Adjusted report messages.
* Removed leapp.db file and added check to .gitignore

After installing newer kernel rpm but without rebooting the host I get the following report entries when running: `snactor run --print-output CheckRebootHygiene`:
```
[
  {
    "stamp": "2023-04-17T11:18:19.391762Z",
    "hostname": "ip-172-31-22-169.eu-central-1.compute.internal",
    "actor": "check_reboot_hygiene",
    "topic": "report_topic",
    "context": "039eeebe-6613-4c95-897a-474da8703609",
    "phase": "NON-WORKFLOW-EXECUTION",
    "message": {
      "hash": "5980fb4c20073e89970f9c3da902421b31ec5e6ed8043e06a4b9e3188f1b8799",
      "data": "{\"report\": \"{\\\"audience\\\": \\\"sysadmin\\\", \\\"detail\\\": {\\\"related_resources\\\": [{\\\"scheme\\\": \\\"default_kernel_version\\\", \\\"title\\\": \\\"3.10.0-1160.88.1.el7.x86_64\\\"}, {\\\"scheme\\\": \\\"running_kernel_version\\\", \\\"title\\\": \\\"3.10.0-1062.1.2.el7.x86_64\\\"}], \\\"remediations\\\": [{\\\"context\\\": \\\"Please reboot the host machine.\\\", \\\"type\\\": \\\"hint\\\"}]}, \\\"flags\\\": [\\\"inhibitor\\\"], \\\"groups\\\": [\\\"sanity\\\", \\\"inhibitor\\\"], \\\"key\\\": \\\"bd8482ac9abc131d9f92443dd874780633278fbc\\\", \\\"severity\\\": \\\"high\\\", \\\"summary\\\": \\\"This host is not running the default kernel version currently configured in the bootloader. Maybe a new kernel version was installed without then rebooting the host? Policy requires the host to be rebooted before going forward with the upgrade.\\\", \\\"title\\\": \\\"Mismatched kernel versions detected\\\"}\"}"
    },
    "type": "Report"
  },
  {
    "stamp": "2023-04-17T11:18:19.405766Z",
    "hostname": "ip-172-31-22-169.eu-central-1.compute.internal",
    "actor": "check_reboot_hygiene",
    "topic": "report_topic",
    "context": "039eeebe-6613-4c95-897a-474da8703609",
    "phase": "NON-WORKFLOW-EXECUTION",
    "message": {
      "hash": "fcfe09723d6c260c48bab63046747e326086750a37a11e6d5eac2de196b44fbd",
      "data": "{\"report\": \"{\\\"audience\\\": \\\"sysadmin\\\", \\\"detail\\\": {\\\"related_resources\\\": [{\\\"scheme\\\": \\\"file\\\", \\\"title\\\": \\\"/boot/initramfs-0-rescue-deac33d2d78647da84fb78a858f8d66f.img\\\"}, {\\\"scheme\\\": \\\"file\\\", \\\"title\\\": \\\"/boot/vmlinuz-0-rescue-deac33d2d78647da84fb78a858f8d66f\\\"}, {\\\"scheme\\\": \\\"file\\\", \\\"title\\\": \\\"/boot/initramfs-3.10.0-1160.88.1.el7.x86_64.img\\\"}, {\\\"scheme\\\": \\\"file\\\", \\\"title\\\": \\\"/boot/initramfs-3.10.0-1062.1.2.el7.x86_64.img\\\"}, {\\\"scheme\\\": \\\"file\\\", \\\"title\\\": \\\"/boot/grub2/grubenv\\\"}, {\\\"scheme\\\": \\\"file\\\", \\\"title\\\": \\\"/boot/grub2/grub.cfg\\\"}, {\\\"scheme\\\": \\\"file\\\", \\\"title\\\": \\\"/boot/grub/grub.conf\\\"}, {\\\"scheme\\\": \\\"file\\\", \\\"title\\\": \\\"/boot/grub/menu.lst\\\"}], \\\"remediations\\\": [{\\\"context\\\": \\\"Please reboot the host machine.\\\", \\\"type\\\": \\\"hint\\\"}]}, \\\"flags\\\": [\\\"inhibitor\\\"], \\\"groups\\\": [\\\"sanity\\\", \\\"inhibitor\\\"], \\\"key\\\": \\\"0e82733798e357f78c42051cb87ab95f199c74f8\\\", \\\"severity\\\": \\\"high\\\", \\\"summary\\\": \\\"Some files on the /boot partition have been modified since the last reboot. To reduce the risk of boot issues related to changes made since the last reboot, policy requires the host to be rebooted before going forward with the upgrade.\\\", \\\"title\\\": \\\"Modified files under /boot detected\\\"}\"}"
    },
    "type": "Report"
  }
]
```

pirat89 commented on repos/system_upgrade/supplements/actors/checkreboothygiene/libraries/checkreboothygiene.py:
```diff
@@ -0,0 +1,126 @@
+        reporting.create_report(report)
+
+
+def check_mismatched_kernel_versions():
+    # Get running kernel version
+    running_kernel_version = _get_booted_kernel()
+
+    # Get default kernel version
+    # NOTE: grubby dependency?
+    try:
+        default_kernel_version = run(["grubby", "--default-kernel"])["stdout"].strip()
+        # Remove the /boot/vmlinuz- prefix from the path
+        default_kernel_version = os.path.basename(default_kernel_version).split("-", 1)[1]
+    except (OSError, CalledProcessError):
+        api.current_logger().warning("Failed to query grubby for default kernel", exc_info=True)
+        return
+
+    if running_kernel_version != default_kernel_version:
+        summary = (
+            "This host is not running the default kernel version currently configured in the bootloader. "
+            "Maybe a new kernel version was installed without then rebooting the host? "
+            "Policy requires the host to be rebooted before going forward with the upgrade."
+        )
+
+        kernel_versions_related = [
+            reporting.RelatedResource("default_kernel_version", default_kernel_version),
+            reporting.RelatedResource("running_kernel_version", running_kernel_version),
+        ]
+
+        report = [
+            reporting.Title("Mismatched kernel versions detected"),
+            reporting.Summary(summary),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Groups([reporting.Groups.SANITY, reporting.Groups.INHIBITOR]),
+            reporting.Remediation(hint="Please reboot the host machine."),
+        ] + kernel_versions_related
+        reporting.create_report(report)
```
@swapdisk What is exact purpose of this check? Is there a benefit to block the upgrade, when you are booted in a newer kernel than the default is set? (the last installed kernel is set always as default, even when an older version (in comparison to the currently booted kernel) is installed as the last one). Currently, the check, whether the system is booted in the newest installed kernel is present for many years. So curious what is the purpose of this one? Also, in case the default kernel is older than the newest one, when this actor is present, the upgrade is always inhibit, as this or our actor block the upgrade. 

The upstream actor: https://github.com/oamg/leapp-repository/tree/master/repos/system_upgrade/common/actors/kernel/checkinstalledkernels

swapdisk replied:
Right, I didn't know there was already a check for booted in the newest installed kernel present for many years. So this check for booted to default kernel version is a little redundant, but still useful as an example of how to make a custom Leapp actor to enforce an organization's policies.

swapdisk commented:
@pirat89 and @Wecros, I'm thinking about writing up a more complete README for the leapp-supplements repo. We could have one main README that would include the details for all the custom actors we ultimately get developed here. Alternatively, the main README could just have a list of the actors and link out to detailed READMEs under the directories of all the different actors. For example, we took this link out approach for the different roles in the Ansible Leapp collection at [oamg/ansible-leapp](https://github.com/oamg/ansible-leapp). Let me know if you have a preference either way?

Wecros commented:
> Let me know if you have a preference either way?

@swapdisk Thank you for doing this! I vote for one big README at this time. I think it might be more readable since there will not be many actors right now. Plus it would probably tie to the process of just looking at the README and adjusting the Makefile vars/targets.

pirat89 commented on repos/system_upgrade_supplements/common/actors/checkreboothygiene/libraries/checkreboothygiene.py:
```diff
@@ -0,0 +1,114 @@
+            "Some files on the /boot partition have been modified since the last reboot. "
+            "To reduce the risk of boot issues related to changes made since the last reboot, "
+            "policy requires the host to be rebooted before going forward with the upgrade."
+            "The following files have been modified:" + FMT_LIST_SEPARATOR.join(sorted_boot_files)
```
@Wecros  current implementation would led to report like this:
```
text text text text. The following files have beenmodified: /boot/whatever/filename
    - /boot/whatever/filename2
...
```

```suggestion
            "The following files have been modified:{}{}"
            .format(FMT_LIST_SEPARATOR, FMT_LIST_SEPARATOR.join(sorted_boot_files))
```
